feat: Support LaTeX and AsciiMath via MathJax on stem.excalidraw.com

DanielJGeiger opened the pull request:
**Notes:**
- Use [#5265](https://github.com/excalidraw/excalidraw/issues/5265) for discussions of [#6037](https://github.com/excalidraw/excalidraw/pull/6037).
- [#6272](https://github.com/excalidraw/excalidraw/issues/6272) lies outside the scope of [#6037](https://github.com/excalidraw/excalidraw/pull/6037).

This PR provides two features: (1) LaTeX/AsciiMath support, and (2) `ExcalidrawElement` subtypes.  The subtypes feature extends the main codebase in order to decouple the implementation of a given subtype (like MathJax) from the rest of Excalidraw.

The `ExcalidrawElement` subtypes feature supports every `ExcalidrawElement` child type.  Another use case besides MathJax: an `ExcalidrawChartElement` could support multiple chart varieties, using `element.subtype = "kroki"`, `element.subtype = "plantuml"`, etc.

#6037 replaces its predecessor #2993.  For details, see #2993, #3915, and #5311.

Fixes #1176, fixes #3569, fixes #5265, fixes #3636, fixes #5865, fixes #5798, fixes #6958.

(Note: #6037 no longer contains the boilerplate for an `@excalidraw/extensions` package.)

DanielJGeiger committed:
feat: Support LaTeX and AsciiMath via MathJax on stem.excalidraw.com

dwelle commented:
On an unrelated note, something I noticed in the previews PR (which is now also part of this one): these completely unrelated features should be made in separate PRs instead of bundled into this one: https://github.com/excalidraw/excalidraw/commit/f5a84f05c39b01b629ad729a5b22592887243676 (https://github.com/excalidraw/excalidraw/issues/5765).

sindzicat commented:
Hello, @DanielJGeiger!
I found, that when I type `\(\begin{bmatrix} a_{11} & a_{12} \\ a_{21} & a_{22} \end{bmatrix}\)` and mode is "text+math", all works fine. But when mode is "math only", then I get two black lines. I suggest it's because that mode "math only" supposes inline equation, and `\begin{bmatrix} ... \end{bmatrix}` and other $\LaTeX$ environments work only for display equations only.

DanielJGeiger committed:
Render LaTeX matrices correctly in math-only mode.

DanielJGeiger committed:
fix: Cache SVGs separately for mixed-text and math-only modes.

DanielJGeiger commented:
@sindzicat Commit 9ee2bf36 should resolve the matrices issue.  Now type `\newline` (instead of `\\`) in LaTeX math-only mode to get linebreaks outside environments like `eqnarray` and `gather`.

Regarding the Unicode spacing issue you reported in https://github.com/excalidraw/excalidraw/pull/2993#issuecomment-1345290674, in which browser(s) did you find that?

sindzicat commented:
@DanielJGeiger, I just tested bmatrix in math only mode, all works fine! Thanks a lot! 👍
As for Unicode spacing, I tested in Brave 1.46.144 (based on Chromium: 108.0.5359.128 (Official Build) (64-bit)) and in Chromium 108.0.5359.125.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
fix: Correct existing subtypes test coverage; add test coverage for
subtype actions; and a subtype action fix.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
fix: Remove leftovers from a merge.

DanielJGeiger committed:
Fix tests introduced by the arrow labels feature.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
Update to latest `Action` changes.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
Filter all context menu items (standard and custom) through
`isActionEnabled`.

DanielJGeiger committed:
Make filtering of custom actions optional.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Fix tests.

r-menezes commented:
This is very useful!
Excalidraw has a huge potential community among students, teachers, professors and professionals that must create mathematical diagrams. This is going to be an awesome addition to the app !

DanielJGeiger committed:
Fixes.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
Simplify custom Actions: universal Action predicates instead of
action-specific guards.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

baptiste commented:
@DanielJGeiger I'm unable to export a page as a shareable link (but Live collaboration works), do you know if this is a limitation of the maths.preview subdomain?
<img width="1073" alt="Screenshot 2023-01-29 at 16 24 37" src="https://user-images.githubusercontent.com/18970/215303037-e7557938-36c3-47d1-95d2-cd6b3aa4a6cf.png">

DanielJGeiger committed:
Refactor: Drop `isActionName` and convert `getCustomActions` to
`filterActions`.

DanielJGeiger commented:
> @DanielJGeiger I'm unable to export a page as a shareable link (but Live collaboration works), do you know if this is a limitation of the maths.preview subdomain?

@baptiste I believe you hit a configuration issue instead ot a limitation.

@dwelle Could you add the `Access-Control-Allow-Origin` CORS header to https://json.excalidraw.com?

DanielJGeiger commented:
> @DanielJGeiger, I just tested bmatrix in math only mode, all works fine! Thanks a lot! +1 As for Unicode spacing, I tested in Brave 1.46.144 (based on Chromium: 108.0.5359.128 (Official Build) (64-bit)) and in Chromium 108.0.5359.125.

@sindzicat I successfully replicated the Unicode spacing issue in Chromium, but not in Firefox.  I have not yet solved it.  But see #6162 for a different situation with browser-specific behaviors.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Refactor: Modify fewer components.

baptiste commented:
@DanielJGeiger have you considered linking this to [Mathpix's OCR API](https://mathpix.com/ocr)? It's not open-source, sadly, but the pricing model seems reasonable enough that it could be of interest. From what I understand, it should be possible to send a SVG path created by doodling in Excalidraw, and have the API return the most likely TeX expression corresponding to the doodle.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/master' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

baptiste commented:
> > @DanielJGeiger I'm unable to export a page as a shareable link (but Live collaboration works), do you know if this is a limitation of the maths.preview subdomain?
> 
> @baptiste I believe you hit a configuration issue instead ot a limitation.
> 
> @dwelle Could you add the `Access-Control-Allow-Origin` CORS header to https://json.excalidraw.com?

@dwelle @ad1992 Would it be possible to allow this, please? I'm currently refraining from showing Excalidraw to students and colleagues, as there's no way to share a link with an existing drawing demonstrating the Mathjax features. This preview branch works really nicely otherwise, as a STEM whiteboard.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
fix: Catch MathML errors and render the "ERR" block instead.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Fixes for math element dimensions before/upon loading MathJax.

mkhan45 commented:
It might be useful to make the background of the math mode textbox a solid color so that it's easier to read when it's expanded over other drawings

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
fix: Better legibility when editing some math elements.

DanielJGeiger committed:
fix: Don't cache wrapped text before MathJax finishes loading.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
fix: Properly avoid concurrent invocations of `loadMathJax()`.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

dwelle commented:
@baptiste @DanielJGeiger loading JSON links from the math subdomain doesn't work for you? E.g. https://math.preview.excalidraw.com/#json=-YxP8jWVTpnrETpH27tl2,Cz0vhRA7Vpe_C0BMUXKCBw

baptiste commented:
> @baptiste @DanielJGeiger loading JSON links from the math subdomain doesn't work for you? E.g. https://math.preview.excalidraw.com/#json=-YxP8jWVTpnrETpH27tl2,Cz0vhRA7Vpe_C0BMUXKCBw

This link works for me, however when I try to generate one I get an error message "Couldn't create shareable link."

![Screenshot 2023-04-09 at 07 30 32](https://user-images.githubusercontent.com/18970/230739459-8f760327-6003-4aa1-8220-9c621fe70a4b.png)

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger commented:
@Cookie04DE To apply AsciiMath vs LaTeX to old text, select each text element you'd like to update and then press *Shift+R*.  (You can also select "Reset math input type" from the canvas context menu.)

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger committed:
Fix a merge lint issue

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

dwelle commented:
Shareable links should now work for the math domain.

DanielJGeiger committed:
fix: No more debounced `refresh()` for subtypes.

DanielJGeiger committed:
Subtypes: add another test.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

diego898 commented:
Just want to bring up another +1 for using `$` instead of `\(` for latex. Jupyter and various other web based uses of mathjax/katex, all use `$` (quarto, stack exchange, etc.)

DanielJGeiger committed:
MathJax: Use `$` as LaTeX delimiters.  Fall back to `\(` and `\)` if detected.  Interpret `\$` as a text literal "$" sign.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

DanielJGeiger commented:
> Just want to bring up another +1 for using `$` instead of `\(` for latex. Jupyter and various other web based uses of mathjax/katex, all use `$` (quarto, stack exchange, etc.)

@diego898 This should work as of commit be057bd.

DanielJGeiger commented:
> When we resize, we observe this peculiar flickering behavior in the equation.
> Screen.Recording.2023-06-06.at.11.32.36.mov

@michelmar As you scale the font size, you're also scaling the exact dimensions in which the MathJax SVG renders to the canvas.  Rather than scaling the font size gradually, try switching back and forth between the predefined font sizes.  Do you see a slight flicker the first time, but not thereafter?

Atreyu-94 commented:
Any advances here? :)

DanielJGeiger commented:
> Any advances here? :)

@Atreyu-94 This PR is just waiting on review by someone from @excalidraw such as @ad1992 or @dwelle.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax

cben commented:
> This [using `$`] should work as of commit be057bde393a479acf343680b525a102a8eea633.

Yep, here is current delimiters behavior (for "Mixed text" blocks):
![excalidraw-math-6037](https://github.com/excalidraw/excalidraw/assets/273688/8d1aec14-7d18-4e37-b030-c265541d947f)

The mode is auto-detected on each line:
* if it contains `\(`, dollars are ignored on that line, only `\(...\)` is supported (and `$` can also be used as-is inside formula);
* otherwise, `$...$` syntax is supported (and inner dollars must be escaped `\$`).

----

"Math only" text blocks take math content without any delimiters.

works great :clap:

DanielJGeiger committed:
fix: Migrate `@excalidraw/extensions` environment variable names to Vite.

DanielJGeiger committed:
chore: Update `@excalidraw/extensions` configs.

DanielJGeiger committed:
@excalidraw/extensions: Fixes for Vite.

DanielJGeiger committed:
fix: Patch AsciiMath to work with Vite.

Incorporates PR mathjax/MathJax-src#854 by @masx200.

DanielJGeiger committed:
fix: Further patch AsciiMath to work with Vite in production mode also.

DanielJGeiger committed:
Add a semicolon.

DanielJGeiger committed:
Revert "Add a semicolon."

This reverts commit 456433e8f0ba5e4bd225151c86ca162e6170c6d6.

DanielJGeiger commented:
@cben You're right, I have not implemented display-math so far.  I'm interested in supporting it, once #6037 is reviewed first.

In addition to your observations about delimiters:  Ordinary dollar signs must be escaped in lines without math.  (`The project would cost \$42 million...`).

DanielJGeiger committed:
chore: Drop `@excalidraw/extensions` and move the MathJax subtype into
`src/excalidraw-app/subtypes` to leave `@excalidraw/excalidraw` untouched.

`@excalidraw/extensions` mostly contained boilerplate and obscured the
main new features here: `ExcalidrawElement` subtypes and MathJax support.

DanielJGeiger committed:
chore: Don't bundle #6050 or #5511.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
refactor: Relocate a type definition.

DanielJGeiger committed:
refactor: Simplify a file.

DanielJGeiger committed:
fix: Inform scenes of mutations when a subtype finishes loading.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

petittitouan commented:
It would be nice if the text areas in "Mixed Mode" could use the handwriting font outside of the $$
(Or add a selector like classic text areas to choose the font for the non-math parts)

DanielJGeiger commented:
> It would be nice if the text areas in "Mixed Mode" could use the handwriting font outside of the $$ (Or add a selector like classic text areas to choose the font for the non-math parts)

@petittitouan This sounds like a worthwhile idea.  However, I cannot realistically implement such requests until my existing MathJax support receives a thorough review.

petittitouan commented:
No problem
Hope you'll get reviewed&merged soon !

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
chore: Only use `transform-origin` in the text editor if rendered
dimensions don't match the editor dimensions.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
fix: Text-only measurements off by a pixel

DanielJGeiger committed:
refactor: Move MathJax into `src/element/subtypes` for the
`excalidraw-app` separation, maintaining lazy-loading of MathJax.

DanielJGeiger committed:
fix: 4d6d6cf1 had a line-height regression for sufficiently short math symbols

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
refactor: Refactor and add a test

DanielJGeiger commented:
Cross-posting from #5265 for reference:

Use #5265 to report any bugs you find while using https://math.preview.excalidraw.com/. Comments posted there are easier to track while #6037 awaits thorough review.
 
Interested parties may also test previews of #6037 at the [deployment](https://excalidraw-git-danieljgeiger-mathjax-maint-stage-excalidraw.vercel.app/) of #6936 before they arrive at https://math.preview.excalidraw.com/.

DanielJGeiger committed:
refactor: Replace the `useSubtypes` selection hook with a generic `useSubtype` hook

juliosilvaebx commented:
@dwelle the size check has failed, yet all sizes are below the configured thresholds. Am I missing something?

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
Refactor the MathJax patch for Vite.

DanielJGeiger committed:
Temporarily drop the MathJax patch.

DanielJGeiger committed:
Re-add the MathJax patch for Vite.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
fix: Don't filter `element.subtype` in `restoreElementWithProperties`.
Subtypes are dynamically registered and may vary across deployments.

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
fix: Only expose `registerAction` instead of `ActionManager` in the API

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
fix: Narrow the type of `Action.name` while still allowing custom names

DanielJGeiger committed:
fix: Distinguish subtype toggles from regular tools.

DanielJGeiger committed:
fix: An empty island would render if no subtypes were registered.

DanielJGeiger committed:
fix: Use a narrower type for custom shortcut names and reduce hard-coding.

DanielJGeiger committed:
feat: Expose `ActionManager.registerAction` through `ExcalidrawImperativeAPI` (#6995)

* feat: Expose `ActionManager` through `ExcalidrawImperativeAPI`

* Only expose `registerAction` instead of `ActionManager`

DanielJGeiger committed:
refactor: Remove some hard-coding in the MathJax subtype.

DanielJGeiger committed:
Resolve merge conflicts

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
chore: Clean up `src/i18n.ts` changes

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
fix: Missed in merging

jiweiyuan reviewed (commented):
The feature is pretty well, thanks

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

DanielJGeiger committed:
Fix test failures from May merge.

baptiste commented:
@DanielJGeiger Have you considered adding [typst](https://typst.app/) support (maybe via [typst.ts](https://myriad-dreamin.github.io/typst.ts/))? It can output SVG natively so I'm guessing it wouldn't be too challenging, given the infrastructure you've already established. One potential advantage is that typst aims for broader font support than TeX and co, so for example a font like https://fonts.nagwa.com/tk/ would be a closer match to Excalidraw's handwritten font.

AAR072 commented:
Any update on this? I really want to use this to tutor remotely but the inability to use latex is holding me back.

bibhas2 commented:
> Any update on this? I really want to use this to tutor remotely but the inability to use latex is holding me back.

Don't let it stop you. The math preview works great. Just click the $\sqrt{x}$ toolbar button and then add formula in a textbox within ``$...$``.

https://math.preview.excalidraw.com/

lmariscal commented:
It appears that the math toggle has disappeared on the latest deployed version of [math.preview.excalidraw.com](https://math.preview.excalidraw.com/)

![Screenshot from 2024-09-03 12-01-17](https://github.com/user-attachments/assets/688627a3-2636-44bc-bf95-707aec71a640)

Or am I missing something?

DanielJGeiger commented:
@lmariscal Vercel was somehow pointing https://math.preview.excalidraw.com to a different branch than this one. But that's fixed now.

psychonaut-s commented:
Is it possible to support colors like so:
`$ \textcolor{red}{x + \delta > y} $`

We need to be able to import the color package like so:

```
MathJax = {
  loader: {load: ['[tex]/color']}, // Load the color package
  tex: {
    packages: {'[+]': ['color']}    // Add the color package to existing ones
  }
};
MathJax.typesetPromise();  // Re-render the content with the new configuration
```

mendax1234 commented:
Is this feature integrated into the official app?

vanchaxy commented:
Not all AsciiMath symbols render on Android. 

Tested on:
* Pixel 7, Android 15, Browsers: Chrome/Firefox
* Samsung Galaxy Tab A7, Android 12, Browsers: Chrome

Correct render at https://asciimath.org/ from phone:
![Correct render at https://asciimath.org/ from phone](https://github.com/user-attachments/assets/eee99f07-33b2-4a39-9d97-fdb11000c7be)

Render input from phone:
![image](https://github.com/user-attachments/assets/c36a9dd5-a56d-439f-948b-5280547201b3)

Render result from phone:
![image](https://github.com/user-attachments/assets/be74f672-fd11-4931-b129-ca76eb6953c2)

Render result from desktop:
![image](https://github.com/user-attachments/assets/0c3bdd53-edc0-408b-a3b7-a3ca1fa0df5d)

albersonmiranda commented:
Really a bummer that this hasn't been merged yet. It's annoying having to step out of VSCode for using the math feature 😭

DanielJGeiger committed:
Merge remote-tracking branch 'origin/release' into danieljgeiger-mathjax-maint-stage

pschyrem commented:
Also waiting for it to be merged.

anchored-archer commented:
Any updates on this one?
